nucleus/utils.py

```diff
@@ -422,3 +435,120 @@ def paginate_generator(
             yield json_value
         if not next_token:
             break
+
+
+def split_s3_bucket_key(s3_path: str):
+    s3_bucket, s3_key = s3_path.split("//", 1)[-1].split("/", 1)
+    return s3_bucket, s3_key
+
+
+def fetch_image(s3_url: str):
+    s3_bucket, s3_key = split_s3_bucket_key(s3_url)
+    image = Image.open(
+        boto3.resource("s3").Bucket(s3_bucket).Object(s3_key).get()["Body"]
+    )
+    return image
+
+
+def fetch_chip(ref_id: str):
+    image_loc = None
+    annotation_loc = None
+    if ref_id.startswith("s3"):
+        s3_bucket, s3_key = split_s3_bucket_key(ref_id)
+        try:
+            boto3.resource("s3").Bucket(s3_bucket).Object(
+                s3_key + ".jpeg"
+            ).load()
+            image_loc = ref_id + ".jpeg"
+        except ClientError:
+            return None, None
+        try:
+            boto3.resource("s3").Bucket(s3_bucket).Object(
+                s3_key + ".json"
+            ).load()
+            annotation_loc = ref_id + ".json"
+        except ClientError:
+            return image_loc, None
+    else:
+        if os.path.exists(ref_id + ".jpeg"):
+            image_loc = ref_id + ".jpeg"
+            if os.path.exists(ref_id + ".json"):
+                annotation_loc = ref_id + ".json"
+    return image_loc, annotation_loc
+
+
+def write_chip(
+    ref_id: str, image: Image.Image, annotations: List[Dict[str, str]]
+):
+    if ref_id.startswith("s3"):
+        s3_bucket, s3_key = split_s3_bucket_key(ref_id)
+        byteio = io.BytesIO()
+        image.save(byteio, format="jpeg")
+        byteio.seek(0)
+        boto3.resource("s3").Bucket(s3_bucket).Object(
+            s3_key + ".jpeg"
+        ).upload_fileobj(byteio)
+        annotation_loc = None
+        if len(annotations) > 0:
+            boto3.resource("s3").Bucket(s3_bucket).Object(
+                s3_key + ".json"
+            ).put(
+                Body=json.dumps(annotations, ensure_ascii=False).encode(
+                    "UTF-8"
+                ),
+                ContentType="application/json",
+            )
+            annotation_loc = ref_id + ".json"
+        return ref_id + ".jpeg", annotation_loc
+    else:
```

Comment:
Is there no danger in treating everything except s3 as local?

Reply:
since we clarify in the doc strings we only support s3 and local users should expect an error if they use a different type